### PR TITLE
DOCSP-28794 AWS Auth Mechanisms

### PR DESCRIPTION
**Pull request by jordan-smith721**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-28794
Staging - https://docs-mongodbcom-staging.corp.mongodb.com/golang/docsworker-xlarge/DOCSP-28794-aws-eks/fundamentals/auth/#mongodb-aws

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?

### CODE REVIEW

File: source/includes/fundamentals/code-snippets/authentication/aws-connection-string-session-token.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+var sessionToken string
```
Comment (qingyang-hu) [inline source/includes/fundamentals/code-snippets/authentication/aws-connection-string-session-token.go:1]: I'm not very familiar with the convention but do we need to list `accessKeyID` as well?

**Review by qingyang-hu — APPROVED** [review]
It look good overall with a side comment on "aws-connection-string-session-token.go".

File: source/fundamentals/auth.txt [hunk 2]
```diff
@@ -189,59 +189,104 @@ username and password to authenticate your user.
...
-- Assign the ``Username`` option the value of your ``accessKeyID``
-- Assign the ``Password`` option the value of your ``secretAccessKey``
+The driver checks for you credentials in the following sources in order:
 
-.. code-block:: go
-   :emphasize-lines: 3, 5-6
+1. Connection string
+#. Environment variables
+#. Web identity token file
+#. AWS ECS endpoint specified in ``AWS_CONTAINER_CREDENTIALS_RELATIVE_URI``
```
Comment (rustagir) [inline source/fundamentals/auth.txt:207]: Q: what is `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`? An option, a property, an object,...?

File: source/fundamentals/auth.txt [hunk 3]
```diff
@@ -189,59 +189,104 @@ username and password to authenticate your user.
...
-      context.TODO(),
-      options.Client().SetAuth(awsCredential))
-   if err != nil {
-      panic(err)
-   }
-   _ = awsIAMClient
+.. important::
+   
+   The driver only reads the credentials from the first method that it detects
+   in the order as given by the preceding list. For example, if you specify
```
Comment (rustagir) [inline source/fundamentals/auth.txt:214]: C: a little confusing

```suggestion
   The driver obtains credentials from the first source that they are detected. For example, if you specify 
```

File: source/fundamentals/auth.txt [hunk 4]
```diff
@@ -189,59 +189,104 @@ username and password to authenticate your user.
...
-      AuthMechanismProperties: map[string]string{
-         "AWS_SESSION_TOKEN": "<sessionToken>",
-      },
-   }
-   assumeRoleClient, err := mongo.Connect(context.TODO(),
-      options.Client().SetAuth(assumeRoleCredential))
+         The following examples set the appropriate credentials using the ``SetAuth()`` 
+         method. You can also specify these credentials using the ``ApplyURI()`` 
+         method. If you use the ``ApplyURI()`` method you must URL encode the username 
+         and password to ensure they are correctly parsed.
```
Comment (rustagir) [inline source/fundamentals/auth.txt:228]: Style guide

```suggestion
         The following examples set the appropriate credentials by using the ``SetAuth()`` 
         method. You can also specify these credentials by using the ``ApplyURI()`` 
         method. If you use the ``ApplyURI()`` method you must URL encode the username 
         and password to ensure they are correctly parsed.
```

File: source/fundamentals/auth.txt [hunk 5]
```diff
@@ -189,59 +189,104 @@ username and password to authenticate your user.
...
+
+      - Assign the ``AuthMechanism`` option the value ``MONGODB-AWS``
+      - Assign the ``Username`` option the value of your ``accessKeyID``
+      - Assign the ``Password`` option the value of your ``secretAccessKey``
+
+      .. literalinclude:: /includes/fundamentals/code-snippets/authentication/aws-connection-string.go
+         :language: go
+
+      If you need to specify an AWS session token, use the temporary
+      credentials returned from an assume role request.
```
Comment (rustagir) [inline source/fundamentals/auth.txt:241]: Q: what is an assume role request?

File: source/fundamentals/auth.txt [hunk 6]
```diff
@@ -189,59 +189,104 @@ username and password to authenticate your user.
...
+      
+      To use temporary credentials, assign the ``AuthMechanismProperties``
+      option the value of your ``sessionToken``:
```
Comment (rustagir) [inline source/fundamentals/auth.txt:244]:
```suggestion
      To use temporary credentials, assign the value of your ``sessionToken`` to the ``AuthMechanismProperties``
      option:
```

File: source/fundamentals/auth.txt [hunk 7]
```diff
@@ -189,59 +189,104 @@ username and password to authenticate your user.
...
+      .. code-block:: bash
+
+         export AWS_ACCESS_KEY_ID=<awsKeyId>
+         export AWS_SECRET_ACCESS_KEY=<awsSecretKey>
+         export AWS_SESSION_TOKEN=<awsSessionToken>
+
+      .. note::
+
+         Omit the line containing ``AWS_SESSION_TOKEN`` if you don't need an AWS
+         session token for that role.
```
Comment (rustagir) [inline source/fundamentals/auth.txt:264]:
```suggestion
         If you don't need an AWS
         session token for that role, omit the line containing ``AWS_SESSION_TOKEN`` .
```

Also, what does "that" refer to in "that role"?

File: source/fundamentals/auth.txt (hunk 2 above)
Comment (jordan-smith721) [reply source/fundamentals/auth.txt]: It's an environment variable - I added that to the bullet

File: source/fundamentals/auth.txt (hunk 3 above)
Comment (jordan-smith721) [reply source/fundamentals/auth.txt]: Adjusted the wording a bit, let me know if that is better

File: source/fundamentals/auth.txt [hunk 8]
```diff
@@ -189,59 +189,104 @@ username and password to authenticate your user.
...
+         method. You can also specify these credentials using the ``ApplyURI()`` 
+         method. If you use the ``ApplyURI()`` method you must URL encode the username 
+         and password to ensure they are correctly parsed.
+   
+      To connect to your MongoDB instance using your AWS IAM credentials, perform the
+      following steps:
+
+      - Assign the ``AuthMechanism`` option the value ``MONGODB-AWS``
+      - Assign the ``Username`` option the value of your ``accessKeyID``
+      - Assign the ``Password`` option the value of your ``secretAccessKey``
```
[the remark this answers is not in the document]
Comment (jordan-smith721) [reply source/fundamentals/auth.txt:236]: No, these don't have to be specified in any specific order, they're just options to set in the object

File: source/fundamentals/auth.txt (hunk 7 above)
Comment (jordan-smith721) [reply source/fundamentals/auth.txt]: Clarified "that" - let me know if it makes more sense now

**Review by rustagir — APPROVED** [review]
lgtm!

**Pull request merged**